### PR TITLE
07960 updating the special files handling for throttles

### PR DESCRIPTION
**Pull request by vtronkov**
**Description**:
* Adding handling for throttle definition files
* Initializing the throttle definitions on genesis(Hedera.java)
* Updating the throttle definition on file update(SystemFileUpdateFacility.java)

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-services/issues/7960

**Notes for reviewer**:
In Hedera, the throttling is configured in a file we've uploaded to our file service. If we want to change that file it needs to go through consensus to be uploaded and most of the nodes need to approve it. 

Currently, we are not reading this file(in the mono code), thus the configuration is not applied. 
In this PR we are creating the `ThrottleManager`. The class is called in two places:
1. `Hedera.genesis` method - when a node is started and there is no previous saved state we are initiating the ThrottleManager by getting the file and saving its content.
2. In the `HandleWorkflow` we are calling the `SystemFileUpdateFacility` which checks if the transaction has a file and if this file is _special_. If it's a _special_ file and it's _throttleDefinitions_ type - we call the `ThrottleManager` to update the content.

The `ThrottleManager` uses the **PROTOBUF** of `ThrottleDefinitions` to parse the bytes and then it saves the list of `ThrottleBucket` as a field in the class. That way when we need to access the throttle definitions we can inject `ThrottleManager`.

**Note:** the sonar coverage is so low because of the changes in Hedera.java class. Currently, there are no tests written for that and it will be very time-consuming if I start writing them from the beginning(I'm guessing we have future stories for that). I've marked what needs to be tested in HederaTest

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleManager.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+     *
+     * @param bytes The protobuf encoded {@link ThrottleDefinitions}.
+     */
+    public void update(@NonNull final Bytes bytes) {
+        // Parse the throttle file. If we cannot parse it, we just continue with whatever our previous rate was.
+        try {
+            throttleDefinitions = ThrottleDefinitions.PROTOBUF.parse(bytes.toReadableSequentialData());
+        } catch (final Exception e) {
+            // Not being able to parse the throttle file is not fatal, and may happen if the throttle file
+            // was too big for a single file update for example.
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleManager.java:58]: Is that true?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/SystemFileUpdateFacility.java [hunk 2]
```diff
@@ -112,7 +116,7 @@ public void handleTxBody(@NonNull final HederaState state, @NonNull final Transa
     }
 
     @NonNull
-    private static Bytes getFileContent(@NonNull final HederaState state, @NonNull final FileID fileID) {
+    public static Bytes getFileContent(@NonNull final HederaState state, @NonNull final FileID fileID) {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/SystemFileUpdateFacility.java:119]: Why was this changed to public?

Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/SystemFileUpdateFacility.java:119]: The only reason I did that is to use it in the `SystemFileUpdateFacilityTest.throttleMangerUpdatedOnFileUpdate` test. If that's not a good enough argument I can make it private again and copy the method into the test class.

Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/SystemFileUpdateFacility.java:119]: **suggestion:** ‏Perhaps make it package private and use `@VisibleForTesting`?

**Review by MiroslavGatsanoga — APPROVED** [review]
LGTM overall, we can create an issue (if there isn't such existing yet) to make the `ThrottleDefValidationSuite` tests pass.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/SystemFileUpdateFacilityTest.java [hunk 3]
```diff
@@ -83,7 +88,7 @@ void testMethodsWithInvalidArguments() {
         final var txBody = simpleCryptoTransfer().body();
 
         // then
-        assertThatThrownBy(() -> new SystemFileUpdateFacility(null)).isInstanceOf(NullPointerException.class);
+        assertThatThrownBy(() -> new SystemFileUpdateFacility(null, null)).isInstanceOf(NullPointerException.class);
```
Comment (netopyr) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/SystemFileUpdateFacilityTest.java:91]: We should have two checks here, one for each parameter.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java [hunk 4]
```diff
@@ -835,6 +842,26 @@ private void initializeExchangeRateManager(@NonNull final HederaState state) {
         logger.info("Exchange rates initialized");
     }
 
+    private void initializeThrottleManager(@NonNull final HederaState state) {
+        logger.info("Initializing throttles");
+        final var readableFileStore = new ReadableStoreFactory(state).getStore(ReadableFileStore.class);
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:847]: There is a lot of code duplicated. Maybe we can extract the common functionality from these initialize...-methods.

**Review by netopyr — APPROVED** [review]
LGTM. Just some minor suggestions which are no blockers.

**Comment by vtronkov** [comment]
@netopyr Since I want to unblock the FE throttling story I will be merging that one and I will address your comments in https://github.com/hashgraph/hedera-services/pull/8147

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java (hunk 4 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:847]: Will fix it in https://github.com/hashgraph/hedera-services/pull/8147

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/SystemFileUpdateFacilityTest.java (hunk 3 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/handle/SystemFileUpdateFacilityTest.java:93]: Will fix it in https://github.com/hashgraph/hedera-services/pull/8147
